metadata/src/main/java/org/apache/kafka/controller/QuorumController.java

```diff
@@ -1293,15 +1358,20 @@ private void handleFeatureControlChange() {
      *
      * @param message           The metadata record
      * @param snapshotId        The snapshotId if this record is from a snapshot
-     * @param batchLastOffset   The offset of the last record in the log batch, or the lastContainedLogOffset
-     *                          if this record is from a snapshot, this is used along with RegisterBrokerRecord
+     * @param offset            The record offset or snapshot offset.
+     * @param epoch             The log epoch of the record or snapshot offset.
      */
-    private void replay(ApiMessage message, Optional<OffsetAndEpoch> snapshotId, long batchLastOffset) {
+    private void replay(
+        ApiMessage message,
+        Optional<OffsetAndEpoch> snapshotId,
+        long offset,
+        int epoch
+    ) {
         logReplayTracker.replay(message);
         MetadataRecordType type = MetadataRecordType.fromId(message.apiKey());
         switch (type) {
             case REGISTER_BROKER_RECORD:
-                clusterControl.replay((RegisterBrokerRecord) message, batchLastOffset);
+                clusterControl.replay((RegisterBrokerRecord) message, offset);
```

Comment:
Currently, we use this offset to remember the offset of each `BrokerRegisterationRecord`, so we can unfence a broker when it's high-watermark has passed it's `BrokerRegisterationRecord`.